### PR TITLE
feat: DiT layer norm fusions for WAN: flashinfer.diffusion_ops

### PR DESCRIPTION
**Pull request by kahyunnam**
## 📌 Description

Add fused DIT (Diffusion Transformer) LayerNorm CUDA kernels for video generation self-attention blocks (WAN 2.2 and similar architectures). Three fused operation modes, each combining multiple steps into a single kernel launch:

| API | Operations Fused |
|-----|-----------------|
| `fused_dit_gate_residual_layernorm_gamma_beta` | `residual_out = residual + input * (gate + gate_bias)` → LayerNorm(gamma, beta) |
| `fused_dit_gate_residual_layernorm_scale_shift` | `residual_out = residual + input * (gate + gate_bias)` → LayerNorm → `(1 + scale + scale_bias) * norm + (shift + shift_bias)` |
| `fused_dit_residual_layernorm_scale_shift` | `residual_out = residual + input` → LayerNorm → `(1 + scale + scale_bias) * norm + (shift + shift_bias)` |

Each API produces two outputs: the residual connection result (BF16) and the normalized output (BF16, NVFP4, or MXFP8). All three are exposed via `flashinfer.diffusion_ops` (like `dsv3_ops`).

### Usage

```python
from flashinfer.diffusion_ops import (
    fused_dit_gate_residual_layernorm_gamma_beta,
    fused_dit_gate_residual_layernorm_scale_shift,
    fused_dit_residual_layernorm_scale_shift,
)

# Self-attention block: gate + residual + LayerNorm(gamma, beta)
residual_out, norm_out = fused_dit_gate_residual_layernorm_gamma_beta(
    input, residual, gate, gamma, beta,
    gate_bias=gate_bias, epsilon=1e-6,
)

# Feed-forward block: gate + residual + LayerNorm + scale/shift (adaLN)
residual_out, norm_out = fused_dit_gate_residual_layernorm_scale_shift(
    input, residual, gate, scale, shift,
    gate_bias=gate_bias, scale_bias=scale_bias, shift_bias=shift_bias,
    epsilon=1e-6,
)

# Cross-attention block: residual + LayerNorm + scale/shift (no gate)
residual_out, norm_out = fused_dit_residual_layernorm_scale_shift(
    input, scale, shift,
    residual=residual, scale_bias=scale_bias, shift_bias=shift_bias,
    epsilon=1e-6,
)
```

All inputs are `[batch, num_rows, 3072]` BF16. Gate/scale/shift accept strided tensors from WAN's `temb.chunk(6, dim=2)` pattern.

### Output Quantization

- **BF16**: Standard output, works on SM80+
- **NVFP4**: 4-bit E2M1 values + E4M3 per-block scale factors (block size 16) with swizzled 128×4 layout. SM100+ only.
- **MXFP8**: 8-bit E4M3 values + E8M0 per-block scale factors (block size 32) with swizzled layout. SM100+ only.

Quantized outputs returned as raw packed tensors + scale factors, matching FlashInfer convention (no torchao dependency).

### Benchmark Results

**B200 (SM100, CUPTI)** — WAN 2.2 5B config (hidden_dim=3072):

| Mode | Shape | Eager (ms) | Fused (ms) | Speedup |
|------|-------|-----------|-----------|---------|
| gate_residual_gamma_beta | bs=1 seq=1920 (production) | 0.134 | 0.017 | **7.8x** |
| gate_residual_gamma_beta | bs=4 seq=1920 | 0.481 | 0.055 | **8.8x** |
| gate_residual_scale_shift | bs=1 seq=1920 (production) | 0.237 | 0.023 | **10.2x** |
| gate_residual_scale_shift | bs=1 seq=768 | 0.147 | 0.011 | **13.5x** |
| residual_scale_shift | bs=1 seq=1920 (production) | 0.186 | 0.018 | **10.6x** |
| residual_scale_shift | bs=1 seq=768 | 0.121 | 0.009 | **14.2x** |

**H100 NVL (SM90):**
- gate_residual_gamma_beta: 7.6–8.3x speedup
- gate_residual_scale_shift: 9.7–10.9x speedup
- residual_scale_shift: 9.8–11.0x speedup

### Architecture Support

| SM | Architecture | BF16 | NVFP4 | MXFP8 |
|----|-------------|------|-------|-------|
| SM80 | Ampere (A100) | Yes | No | No |
| SM89 | Ada (L40S) | Yes | No | No |
| SM90 | Hopper (H100) | Yes (primary target) | No | No |
| SM100 | Blackwell (B200) | Yes (primary target) | Yes | Yes |
| SM103 | Blackwell (B300) | Yes (primary target) | Yes | Yes |
| SM110–SM121 | Blackwell variants | Yes | Expected | Expected |

float2 packed math intrinsics (`__fadd2_rn`, `__fmul2_rn`, `__ffma2_rn`) use native SM100+ instructions with scalar fallbacks for SM80/89/90.

### Known Limitations

- **hidden_dim = 3072 only**: Hardcoded block_size=384 with elem_per_thread=4. WAN 1.3B (1536) and 14B (5120) unsupported.
- **Gate/scale/shift stride = 6**: Kernel expects tensors from WAN's `temb.chunk(6, dim=2)` pattern with stride `6 * hidden_dim` in the row dimension. Documented and validated at API level.
- **BF16 input only**: No FP16/FP32 input support.

### Bugs Found and Fixed During Integration

- **Infinite recursion on SM100+**: `f2math::fadd2` wrapper called itself instead of `::__fadd2_rn`. Fixed by using global namespace `::` prefix.
- **Missing `cuda_fp16.h` include**: float2 math intrinsic declarations required this header.
- **Variance NaN safety**: Added `max(0, variance)` clamp before `frsqrt` to prevent NaN from floating-point cancellation in `E[x²] - (E[x])²`.

### Files Changed

```
include/flashinfer/norm/fused_dit_layernorm.cuh      # CUDA kernel + utilities (887 lines)
csrc/norm.cu                                           # TVM-FFI launcher (added to norm module)
csrc/flashinfer_norm_binding.cu                        # TVM-FFI export (added to norm module)
flashinfer/norm/__init__.py                            # 3 Python API functions + helpers
flashinfer/diffusion_ops/__init__.py                   # User-facing facade (like dsv3_ops/)
tests/norm/test_fused_dit_layernorm.py                 # 34 tests
benchmarks/bench_fused_dit_layernorm.py                # Standalone benchmark
benchmarks/routines/norm.py                            # Unified benchmark CLI integration
benchmarks/routines/flashinfer_benchmark_utils.py      # Routine registration
```

## 🔍 Related Issues

https://github.com/flashinfer-ai/flashinfer/issues/2970

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit`.
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing.

**Test summary: 34 tests**

| GPU | Passed | Skipped |
|-----|--------|---------|
| A100 (SM80) | 25 | 9 |
| L40S (SM89) | 25 | 9 |
| H100 (SM90) | 25 | 9 |
| **B200 (SM100)** | **34** | **0** |

Tests cover:
- 3 modes × 5 shapes = 15 BF16 correctness tests
- 2 destination-passing tests (gamma_beta + scale_shift modes)
- residual=None (no residual addition)
- Odd num_rows with WAN-style strided inputs
- 2 modes × 2 shapes = 4 NVFP4 output tests (SM100+ only)
- 2 modes × 2 shapes = 4 MXFP8 output tests (SM100+ only)
- 7 error-case validation tests

## Reviewer Notes

- The kernel compiles as part of the existing `norm` JIT module — no new JIT spec or AOT registration needed.
- Gate/scale/shift tensors use WAN's `temb.chunk(6, dim=2)` stride convention (stride=6 in row dim). This is documented in all three API docstrings.
- The `diffusion_ops/` facade follows the `dsv3_ops/` pattern — pure re-export, no implementation.
- All three API functions support optional pre-allocated outputs (destination-passing style) with shape/dtype validation.
- NVFP4/MXFP8 outputs are returned as raw packed tensors + scale factors (no torchao dependency), matching FlashInfer convention.

## Summary by CodeRabbit

* **New Features**
  * Added three fused DiT-style LayerNorm APIs (gate+residual variants and residual-only) with BF16 and new quantized outputs (NVFP4, MXFP8), input/shape/SM validations, and optional destination buffers.

* **Benchmarks**
  * New GPU benchmark script comparing fused kernels vs eager reference across WAN-like batch/sequence shapes and three modes, reporting median timings and speedups.

* **Tests**
  * New comprehensive tests for correctness, quantized outputs, destination-passing, odd lengths, edge cases, and validation errors.

### CODE REVIEW

File: flashinfer/norm/__init__.py [hunk 1]
```diff
@@ -761,6 +761,651 @@ def fused_rmsnorm_silu(
...
+            if global_scaling_factor is not None
+            else torch.ones(1, dtype=torch.float32, device=device)
+        )
+    elif output_format == _DIT_LN_OUTPUT_MXFP8:
+        expected_norm_shape = (batch_size, num_rows, hidden_size // 4)
+        expected_norm_dtype = torch.int32
+        numMTiles = (num_rows + 127) // 128
+        numKTiles = (hidden_size + 127) // 128
+        expected_sf_shape = (batch_size, numMTiles, numKTiles, 32, 4, 4)
+        sf_scale = torch.zeros(1, dtype=torch.float32, device=device)
```
[the remark this answers is not in the document]
Comment (kahyunnam) [reply flashinfer/norm/__init__.py:1001]: SFScaleVal is only in FP4Converter (lines 303-364). The FP8Converter (starting around line 392) doesn't use sf_scale at all — it computes E8M0 scale factors directly from the data

**Review by bkryu — COMMENTED** [review]
@kahyunnam, the code bits themselves generally look okay to me but question:

Do you consider the newly added operations as "DiT-specific optimizations" or "norm fusion variants"? I know the boundary is not clear-cut, but asking because we are creating a directory `diffusion_ops` and placing an init file in there, but the actual kernel itself is placed in `include/flashinfer/norm` and the associated Python code and tests are also found in `flashinfer/norm/__init__.py` or `tests/norm/test_fused_dit_layernorm.py`

Other existing norm fusion variants such as add + norm + quant variants are generic enough to be considered a norm variant. However, the current organization feels a bit confusing. If we see the new operations as being solely (or even primarily) by diffusion transformers, I'd naively think it might be best to place them in diffusion_ops (with new directories for the kernel file and test files).

I'm curious what you think

**Comment by kahyunnam** [comment]
@bkryu the current set up is somewhat of a compromise between them; right now, these APIs are technically accessible from both `flashinfer.diffusion_ops` module and `flashinfer.norm` module. If I had to pick one, though, I think these are more aligned with `flashinfer.diffusion_ops` since these operations only work specifically for hidden_dim=3072 (WAN 2.2 5B config).

The `flashinfer.dsv3_ops` module also follows a similar idea of each op being from the op module, and just imported into the model specific folder: https://github.com/flashinfer-ai/flashinfer/blob/main/flashinfer/dsv3_ops/__init__.py 

But then `flashinfer.gdn_kernels` module defines the gdn kernels (technically attention?) directly in the directory. Meanwhile, llama-specific RoPE APIs are only defined in the `flashinfer.rope` module, and just named with "llama" in the API name ...

**Review by bkryu — APPROVED** [review]
